src/azure/storage/blob/blob_stream_builder.rs

```diff
@@ -0,0 +1,409 @@
+            lease_id: self.lease_id,
+            client_request_id: self.client_request_id,
+            increment: self.increment,
+        }
+    }
+}
+
+impl<'a, ContainerNameSet, BlobNameSet, RangeSet> LeaseIdSupport<'a> for BlobStreamBuilder<'a, ContainerNameSet, BlobNameSet, RangeSet>
+where
+    ContainerNameSet: ToAssign,
+    BlobNameSet: ToAssign,
+    RangeSet: ToAssign,
+{
+    type O = BlobStreamBuilder<'a, ContainerNameSet, BlobNameSet, RangeSet>;
+
+    #[inline]
+    fn with_lease_id(self, lease_id: &'a LeaseId) -> Self::O {
+        BlobStreamBuilder {
+            client: self.client,
+            p_container_name: PhantomData {},
+            p_blob_name: PhantomData {},
+            p_range: PhantomData {},
+            container_name: self.container_name,
+            blob_name: self.blob_name,
+            snapshot: self.snapshot,
+            timeout: self.timeout,
+            range: self.range,
+            lease_id: Some(lease_id),
+            client_request_id: self.client_request_id,
+            increment: self.increment,
+        }
+    }
+}
+
+impl<'a, ContainerNameSet, BlobNameSet, RangeSet> ClientRequestIdSupport<'a>
+    for BlobStreamBuilder<'a, ContainerNameSet, BlobNameSet, RangeSet>
+where
+    ContainerNameSet: ToAssign,
+    BlobNameSet: ToAssign,
+    RangeSet: ToAssign,
+{
+    type O = BlobStreamBuilder<'a, ContainerNameSet, BlobNameSet, RangeSet>;
+
+    #[inline]
+    fn with_client_request_id(self, client_request_id: &'a str) -> Self::O {
+        BlobStreamBuilder {
+            client: self.client,
+            p_container_name: PhantomData {},
+            p_blob_name: PhantomData {},
+            p_range: PhantomData {},
+            container_name: self.container_name,
+            blob_name: self.blob_name,
+            snapshot: self.snapshot,
+            timeout: self.timeout,
+            range: self.range,
+            lease_id: self.lease_id,
+            client_request_id: Some(client_request_id),
+            increment: self.increment,
+        }
+    }
+}
+
+// methods callable regardless
+impl<'a, ContainerNameSet, BlobNameSet, RangeSet> BlobStreamBuilder<'a, ContainerNameSet, BlobNameSet, RangeSet>
+where
+    ContainerNameSet: ToAssign,
+    BlobNameSet: ToAssign,
+    RangeSet: ToAssign,
+{}
+
+impl<'a> BlobStreamBuilder<'a, Yes, Yes, Yes> {
+    #[inline]
+    pub fn finalize(self) -> impl Stream<Item = Vec<u8>, Error = AzureError> {
+        let container_name = self.container_name().to_owned();
+        let blob_name = self.blob_name().to_owned();
+        let range = self.range().to_owned();
+
+        let snapshot = self.snapshot.to_owned();
+        let timeout = self.timeout.to_owned();
+        //let lease_id = self.lease_id.to_owned();
+        //let client_request_id = self.client_request_id.to_owned();
+        let increment = self.increment;
+
+        let client = self.client().clone();
+
+        stream::unfold(Some(range), move |remaining| {
```

Comment:
This is really clever!

Reply:
Thanks but credits go to the blogpost linked in the list stream pr. (I'm planning the rework that pr to have the same design as here: keep the existing future and add a streaming version based on that